rust/agama-network/src/dbus/interfaces.rs

```diff
@@ -0,0 +1,402 @@
+//! Network D-Bus interfaces.
+//!
+//! This module contains the set of D-Bus interfaces that are exposed by [D-Bus network
+//! service](crate::NetworkService).
+use super::ObjectsRegistry;
+use crate::{
+    action::Action,
+    error::NetworkStateError,
+    model::{Connection as NetworkConnection, Device as NetworkDevice, WirelessConnection},
+};
+use parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
+use std::{
+    net::{AddrParseError, Ipv4Addr},
+    sync::{mpsc::Sender, Arc},
+};
+use uuid::Uuid;
+use zbus::{dbus_interface, zvariant::ObjectPath};
+
+/// D-Bus interface for the network devices collection
+///
+/// It offers an API to query the devices collection.
+pub struct Devices {
+    objects: Arc<Mutex<ObjectsRegistry>>,
+}
+
+impl Devices {
+    /// Creates a Devices interface object.
+    ///
+    /// * `objects`: Objects paths registry.
+    pub fn new(objects: Arc<Mutex<ObjectsRegistry>>) -> Self {
+        Self { objects }
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Devices")]
+impl Devices {
+    /// Returns the D-Bus paths of the network devices.
+    pub fn get_devices(&self) -> Vec<ObjectPath> {
+        let objects = self.objects.lock();
+        objects
+            .devices_paths()
+            .iter()
+            .filter_map(|c| ObjectPath::try_from(c.clone()).ok())
```

Comment:
so if I get it, it silently ignore if  there is some wrong device paths? it should contain at least explanation why object can fail to be converted to object path

Reply:
Yes, good point, we should report that somehow.

Reply:
In a strongly typed language I would expect the object paths to be always valid, unless coming as user input or as arguments to a constructor, which is not the case here.